Release/7.1 Eventing Listening to Multiple Collections (#2851)
couchbase/docs-server: modules/eventing/pages/eventing-advanced-keyspace-accessors.adoc

For example rather than blindly relying on he Basic Keyspace Accessors for an upsert like operation ‘src_col[id_str] = some_doc’ the Advanced Accessors allow you to resolve contention (or possible contention) with JavaScript driven logic in your Eventing Function.

* If the document doesn't exist you can use ‘couchbase.insert(src_col, {"id: id_str}, some_doc)’ and check the return value for success

> * If you don't care if the document exists you can use ‘couchbase.upsert(src_col, {"id: id_str}, some_doc)’ and check the return value for success

Reviewer: ```suggestion
* You can use ‘couchbase.upsert(src_col, {"id: id_str}, some_doc)’ to update a document if it already exists. 
If the document doesn't exist, then a new document will be created. 
Check the return value to see if the operation was successful.
```

* If the document exists you can use ‘couchbase.replace(src_col, {"id: id_str, "cas": current_cas}, some_doc)’ and check the return value for success or a CAS mismatch.

To view complete examples including JavaScript, input mutations, output mutations and/or log messages for each Advanced Bucket Accessor refer to 
xref:eventing-examples.adoc#examples-scriptlets-advanced-accessors[Scriptlets: Advanced Accessor Handlers] in the examples section of the documentation.

*Functions that Listen to Multiple Collections*: may require the _meta.keyspace_ parameter [.status]#Couchbase Server 7.1.1#

> For Eventing Functions that listen to multiple collections via a wildcard of "+++*+++" for the scope and/or the collection an additional parameter is be required if the binding used by the Advanced Keyspace Accessor also contains a wild card of "+++*+++" for the scope and/or the collection.

Reviewer: ```suggestion
For Eventing Functions that listen to multiple collections via a wildcard of `+++*+++` for the scope and/or the collection, an additional parameter is required if the binding used by the Advanced Keyspace Accessor also contains a wild card of `+++*+++` for the scope and/or the collection.
```

* ‘couchbase.insert(src_col, {"id: id_str, "keyspace":{"bucket_name": "bkt01","scope_name":"scp01","collection_name":"col01"}}, some_doc)’ 

> The example above adds a _keyspace_ to the _meta_ object to specify win which keyspace the insert is to take place.  Note as of 7.1.1 the entry points of OnUpdate() and OnDelete() provide _meta.keyspace_ with every mutation.  For a detailed example refer to the multiCollectionEventing example in xref:eventing-examples.adoc#examples-scriptlets-advanced-accessors[Scriptlets: Advanced Accessor Handlers] 

Reviewer: ```suggestion
The example above adds a _keyspace_ to the _meta_ object to specify win which keyspace the insert is to take place.  
Note as of 7.1.1 the entry points of OnUpdate() and OnDelete() provide _meta.keyspace_ with every mutation.  
For a detailed example refer to the multiCollectionEventing example in xref:eventing-examples.adoc#examples-scriptlets-advanced-accessors[Scriptlets: Advanced Accessor Handlers] 
```

+
The key of the document to be used for the operation. This is a mandatory parameter and must be of JavaScript _string_ type.

** _meta.keyspace (type: Object)_
+

> The keyspace of the document to be used for the operation with the JavaScript format "keyspace":{"bucket_name":_string_,"scope_name":_string_,"collection_name":_string_}. This parameter is only required if you are using a a _binding_ with a wildcard for the scope and/or the collection. [.status]#Couchbase Server 7.1.1#

Reviewer: ```suggestion
The keyspace of the document to be used for the operation with the JavaScript format "keyspace":{"bucket_name":_string_,"scope_name":_string_,"collection_name":_string_}. 
This parameter is only required if you are using a _binding_ with a wildcard for the scope and/or the collection. [.status]#Couchbase Server 7.1.1#
```

+
The key of the document to be used for the operation. This is a mandatory parameter and must be of JavaScript _string_ type. If a document already exists with the specified key, the operation fails and _key_already_exists_ will be set in the error object.

** _meta.keyspace (type: Object)_
+

> The keyspace of the document to be used for the operation with the JavaScript format "keyspace":{"bucket_name":_string_,"scope_name":_string_,"collection_name":_string_}. This parameter is only required if you are using a a _binding_ with a wildcard for the scope and/or the collection. [.status]#Couchbase Server 7.1.1#

Reviewer: ```suggestion
The keyspace of the document to be used for the operation with the JavaScript format "keyspace":{"bucket_name":_string_,"scope_name":_string_,"collection_name":_string_}. 
This parameter is only required if you are using  a _binding_ with a wildcard for the scope and/or the collection. [.status]#Couchbase Server 7.1.1#
```

+

> The key of the document to be used for the operation. This is a mandatory parameter and must be of JavaScript _string_ type.

Reviewer: Is all this supposed to be in bold?

** _meta.keyspace (type: Object)_
+

> The keyspace of the document to be used for the operation with the JavaScript format "keyspace":{"bucket_name":_string_,"scope_name":_string_,"collection_name":_string_}. This parameter is only required if you are using a a _binding_ with a wildcard for the scope and/or the collection. [.status]#Couchbase Server 7.1.1#

Reviewer: ```suggestion
The keyspace of the document to be used for the operation with the JavaScript format "keyspace":{"bucket_name":_string_,"scope_name":_string_,"collection_name":_string_}. 
This parameter is only required if you are using a _binding_ with a wildcard for the scope and/or the collection. [.status]#Couchbase Server 7.1.1#
```

+
The key of the document to be used for the operation. This is a mandatory parameter and must be of JavaScript _string_ type. If the specified key is not present in the bucket, the operation fails and _key_not_found_ will be set in the error object.

** _meta.keyspace (type: Object)_
+

> The keyspace of the document to be used for the operation with the JavaScript format "keyspace":{"bucket_name":_string_,"scope_name":_string_,"collection_name":_string_}. This parameter is only required if you are using a a _binding_ with a wildcard for the scope and/or the collection. [.status]#Couchbase Server 7.1.1#

Reviewer: ```suggestion
The keyspace of the document to be used for the operation with the JavaScript format "keyspace":{"bucket_name":_string_,"scope_name":_string_,"collection_name":_string_}. 
This parameter is only required if you are using  a _binding_ with a wildcard for the scope and/or the collection. [.status]#Couchbase Server 7.1.1#
```

As a general comment, it's usually a good idea to have repeated text set up as an Asciidoc attribute, instead of copying and pasting.

+
The key of the document to be used for the operation. This is a mandatory parameter and must be of JavaScript _string_ type. If the specified key is not present in the bucket, the operation fails and _key_not_found_ will be set in the error object.

** _meta.keyspace (type: Object)_
+

> The keyspace of the document to be used for the operation with the JavaScript format "keyspace":{"bucket_name":_string_,"scope_name":_string_,"collection_name":_string_}. This parameter is only required if you are using a a _binding_ with a wildcard for the scope and/or the collection. [.status]#Couchbase Server 7.1.1#

Reviewer: ```suggestion
The keyspace of the document to be used for the operation with the JavaScript format "keyspace":{"bucket_name":_string_,"scope_name":_string_,"collection_name":_string_}. 
This parameter is only required if you are using a _binding_ with a wildcard for the scope and/or the collection. [.status]#Couchbase Server 7.1.1#
```

+
The key of the document to be used for the operation. This is a mandatory parameter and must be of JavaScript _string_ type.

** _meta.keyspace (type: Object)_
+

> The keyspace of the document to be used for the operation with the JavaScript format "keyspace":{"bucket_name":_string_,"scope_name":_string_,"collection_name":_string_}. This parameter is only required if you are using a a _binding_ with a wildcard for the scope and/or the collection. [.status]#Couchbase Server 7.1.1#

Reviewer: ```suggestion
The keyspace of the document to be used for the operation with the JavaScript format "keyspace":{"bucket_name":_string_,"scope_name":_string_,"collection_name":_string_}. 
This parameter is only required if you are using a  _binding_ with a wildcard for the scope and/or the collection. [.status]#Couchbase Server 7.1.1#
```

+
The key of the document to be used for the operation. This is a mandatory parameter and must be of JavaScript _string_ type.

** _meta.keyspace (type: Object)_
+

> The keyspace of the document to be used for the operation with the JavaScript format "keyspace":{"bucket_name":_string_,"scope_name":_string_,"collection_name":_string_}. This parameter is only required if you are using a a _binding_ with a wildcard for the scope and/or the collection. [.status]#Couchbase Server 7.1.1#

Reviewer: ```suggestion
The keyspace of the document to be used for the operation with the JavaScript format "keyspace":{"bucket_name":_string_,"scope_name":_string_,"collection_name":_string_}. 
This parameter is only required if you are using a _binding_ with a wildcard for the scope and/or the collection. [.status]#Couchbase Server 7.1.1#
```

Yes, all these repeated blocks have the same error, `a a _binding_,`, so it's definitely worth using an attribute or an included file for the common text if possible. (That way the error only needs correcting in one place).
